File: scripts/dytranscript.js
/**
 * dytranscript — Node.js 版本
 * 通过 CDP 连接已登录的 Chrome → 拦截抖音 API → 提取 video_text 逐字稿
 *
 * 第1步：关掉所有 Chrome，Win+R 执行:
 *   chrome.exe --remote-debugging-port=9222
 * 第2步：在打开的 Chrome 中登录 douyin.com
 * 第3步：运行本脚本
 *
 * 用法: node dytranscript.js <视频链接> [--batch] [--output-dir=<目录>]
 *   node dytranscript.js https://www.douyin.com/video/7643746305843514675
 *   node dytranscript.js --batch --input=../douyin_videos.json
 */
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');
const https = require('http');

// ===== 配置 =====
const CDP_PORT = 9222;
const OUTPUT_DIR = path.join(require('os').homedir(), '.dytranscript_output');
if (!fs.existsSync(OUTPUT_DIR)) fs.mkdirSync(OUTPUT_DIR, { recursive: true });

// ===== 工具函数 =====
function p(msg) {
  try { console.log(msg); } catch (e) { console.log(String(msg)); }
}

function extractVideoId(text) {
  let m = text.match(/douyin\.com\/video\/(\d+)/);
  if (m) return m[1];
  m = text.match(/v\.douyin\.com\/([a-zA-Z0-9_-]+)/);
  if (m) return `short:${m[1]}`;
  return null;
}

function fmtTs(s) {
  const m = Math.floor(parseInt(s) / 60);
  const sec = parseInt(s) % 60;
  return `${String(m).padStart(2, '0')}:${String(sec).padStart(2, '0')}`;
}

function extractTranscript(detail) {
  const result = {
    author: (detail.author || {}).nickname || '',
    desc: detail.desc || '',
    create_time: '',
    stats: {},
    hashtags: [],
    video_text: [],
    has_captions: false,
    aweme_id: detail.aweme_id || '',
  };

  const st = detail.statistics || {};
  result.stats = {
    likes: st.digg_count || 0,
    comments: st.comment_count || 0,
    favorites: st.collect_count || 0,
    shares: st.share_count || 0,
    plays: st.play_count || 0,
  };

  const ct = detail.create_time;
  if (ct) result.create_time = new Date(ct * 1000).toISOString().replace('T', ' ').substring(0, 16);

  for (const te of (detail.text_extra || [])) {
    if (te.hashtag_name) result.hashtags.push(te.hashtag_name);
  }

  const vt = detail.video_text || [];
  if (vt.length > 0) {
    result.has_captions = true;
    for (const seg of vt) {
      result.video_text.push({
        text: seg.text || '',
        start_time: seg.start_time || 0,
        end_time: seg.end_time || 0,
      });
    }
  }

  return result;
}

function outputTranscript(data, videoUrl) {
  p('='.repeat(55));
  p('🎬 抖音文案提取完成');
  p('='.repeat(55));
  if (data.author) p(`👤 ${data.author}`);
  if (data.stats) {
    const s = data.stats;
    p(`📊 👍${s.likes}  💬${s.comments}  ⭐${s.favorites}`);
  }
  if (data.hashtags.length) p(`🏷  ${data.hashtags.map(h => '#' + h).join('  ')}`);
  if (data.create_time) p(`🕐 ${data.create_time}`);
  p(`🔗 ${videoUrl}`);

  const desc = data.desc.trim();
  if (desc) p(`\n📝 文案描述:\n${'-'.repeat(40)}\n${desc}`);

  const vt = data.video_text;
  if (vt.length > 0) {
    p(`\n🎤 逐字稿 (${vt.length} 条):\n${'-'.repeat(40)}`);
    const fullText = [];
    for (const seg of vt) {
      p(`  [${fmtTs(seg.start_time)}] ${seg.text}`);
      fullText.push(seg.text);
    }
    p(`${'-'.repeat(40)}\n\n📋 全文:\n${fullText.join('')}`);
  } else if (!data.has_captions) {
    p('\n💡 该视频没有自动字幕，只有文案描述。');
  }
  p('='.repeat(55));

  // Save to file
  const ts = new Date().toISOString().replace(/[:.]/g, '').substring(0, 15);
  const name = (data.author || 'dy').replace(/[\\/:*?"<>|]/g, '').substring(0, 16);
  const fp = path.join(OUTPUT_DIR, `dy_${name}_${ts}.json`);
  fs.writeFileSync(fp, JSON.stringify(data, null, 2), 'utf-8');
  p(`\n💾 已保存: ${fp}`);
  return data;
}

// ===== CDP 连接检查 =====
async function checkCDP() {
  return new Promise((resolve) => {
    const req = https.get(`http://127.0.0.1:${CDP_PORT}/json/version`, (res) => {
      let data = '';
      res.on('data', c => data += c);
      res.on('end', () => {
        try {
          const d = JSON.parse(data);
          resolve(d.Browser ? d.Browser : null);
        } catch (e) { resolve(null); }
      });
    });
    req.on('error', () => resolve(null));
    req.setTimeout(3000, () => { req.destroy(); resolve(null); });
  });
}

// ===== 主流程: 提取单个视频 =====
async function extractVideo(videoIdOrUrl) {
  const videoId = extractVideoId(videoIdOrUrl);
  if (!videoId) {
    p('❌ 未识别到抖音视频链接');
    return null;
  }

  // Resolve short URL
  let resolvedId = videoId.startsWith('short:') ? videoId.split(':')[1] : videoId;
  let videoUrl = videoId.startsWith('short:')
    ? `https://v.douyin.com/${resolvedId}/`
    : `https://www.douyin.com/video/${videoId}`;

  // Check CDP
  const browserInfo = await checkCDP();
  if (!browserInfo) {
    p('\n' + '!'.repeat(50));
    p('需要先以调试模式启动 Chrome');
    p('!'.repeat(50));
    p('\n请按以下步骤操作：');
    p('  ① 完全关闭所有 Chrome 窗口');
    p('  ② 按 Win+R，粘贴并回车：');
    p('     chrome.exe --remote-debugging-port=9222');
    p('  ③ 在打开的 Chrome 中登录抖音（如未登录）');
    p('  ④ 重新运行即可');
    return null;
  }

  // Connect via CDP
  const browser = await chromium.connectOverCDP(`http://127.0.0.1:${CDP_PORT}`);
  const ctx = browser.contexts()[0] || (await browser.newContext());
  const page = ctx.pages()[0] || (await ctx.newPage());

  p(`\n✅ 已连接到 ${browserInfo}`);

  // Check login
  p('检查登录状态...');
  await page.goto('https://www.douyin.com/', { waitUntil: 'domcontentloaded', timeout: 20000 });
  await page.waitForTimeout(3000);
  const cookies = await ctx.cookies();
  const loggedIn = cookies.some(c =>
    ['sessionid', 'sid_guard', 'sid_tt'].includes(c.name) && c.value
  );
  if (!loggedIn) {
    p('❌ 未登录抖音，请在 Chrome 中登录后重试');
    return null;
  }
  p('✅ 已登录');

  // Resolve short URL if needed
  if (videoId.startsWith('short:')) {
    const np = await ctx.newPage();
    await np.goto(videoUrl, { waitUntil: 'domcontentloaded', timeout: 30000 });
    await np.waitForTimeout(2000);
    const m = np.url().match(/douyin\.com\/video\/(\d+)/);
    if (m) {
      resolvedId = m[1];
      videoUrl = np.url();
      p(`  → 解析为完整ID: ${resolvedId}`);
    }
    await np.close();
  }

  // Capture detail API response
  let detailData = null;

  page.on('response', async (resp) => {
    const url = resp.url();
    if (url.includes('aweme/v1/web/aweme/detail/')) {
      try {
        const data = await resp.json();
        if (data.aweme_detail) {
          detailData = data.aweme_detail;
        }
      } catch (e) {}
    }
  });

  // Navigate to video page
  p('打开视频页面...');
  await page.goto(videoUrl, { waitUntil: 'domcontentloaded', timeout: 30000 });

  // Wait for detail data
  for (let i = 0; i < 20; i++) {
    await page.waitForTimeout(1000);
    if (detailData) break;
  }
  await page.waitForTimeout(3000);

  if (!detailData) {
    p('⚠️ 未获取到视频数据（可能需要登录或视频不存在）');
    return null;
  }

  const data = extractTranscript(detailData);
  outputTranscript(data, videoUrl);
  return data;
}

// ===== 批量处理 =====
async function batchExtract(videoList) {
  const results = [];
  const videos = JSON.parse(fs.readFileSync(videoList, 'utf-8'));
  const total = videos.length;

  p(`\n📋 批量处理 ${total} 个视频\n`);

  for (let i = 0; i < total; i++) {
    const v = videos[i];
    p(`\n[${i + 1}/${total}] ${v.desc?.substring(0, 60) || v.share_url}`);
    try {
      const result = await extractVideo(v.share_url);
      results.push(result);
    } catch (e) {
      p(`  ❌ 失败: ${e.message}`);
      results.push({ aweme_id: v.aweme_id, share_url: v.share_url, error: e.message });
    }

    // Save incremental
    fs.writeFileSync(
      path.join(OUTPUT_DIR, 'batch_results.json'),
      JSON.stringify(results, null, 2),
      'utf-8'
    );

    // Rate limit
    await new Promise(r => setTimeout(r, 3000));
  }

  p(`\n✅ 批量完成: ${results.filter(r => r && r.has_captions).length}/${total} 有逐字稿`);
  fs.writeFileSync(path.join(OUTPUT_DIR, 'batch_results.json'), JSON.stringify(results, null, 2), 'utf-8');
  return results;
}

// ===== 入口 =====
async function main() {
  const args = process.argv.slice(2);

  if (args.includes('--batch')) {
    const inputArg = args.find(a => a.startsWith('--input='));
    if (!inputArg) {
      p('用法: node dytranscript.js --batch --input=<json文件路径>');
      process.exit(1);
    }
    await batchExtract(inputArg.split('=')[1]);
    return;
  }

  if (args.length < 1 || args[0].startsWith('--')) {
    p('用法:');
    p('  单视频: node dytranscript.js <抖音视频链接>');
    p('  批量:   node dytranscript.js --batch --input=videos.json');
    p('\n首次使用:');
    p('  1. 关闭所有 Chrome');
    p('  2. Win+R 输入: chrome.exe --remote-debugging-port=9222');
    p('  3. 登录抖音');
    p('  4. 运行本脚本');
    return;
  }

  // Find douyin URL in all args
  const input = args.join(' ');
  const videoId = extractVideoId(input);
  if (!videoId) {
    const linkInArgs = args.find(a => a.includes('douyin.com') || a.includes('v.douyin.com'));
    if (linkInArgs) {
      await extractVideo(linkInArgs);
    } else {
      p('❌ 未识别到抖音链接。请传入完整的 douyin.com/video/xxx 链接');
    }
    return;
  }

  await extractVideo(input);
}

main().catch(e => {
  p(`\n❌ 错误: ${e.message}`);
  console.error(e);
  process.exit(1);
});

File: scripts/dytranscript.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
抖音逐字稿提取器 — CDP 连接模式

用法:
  第1步：关掉所有 Chrome 窗口
  第2步：在「运行」(Win+R) 粘贴下面这行并回车：
    chrome.exe --remote-debugging-port=9222

  第3步：在打开的 Chrome 中登录抖音（如果没登录）
  第4步：运行本脚本：
    python dytranscript.py "https://www.douyin.com/video/xxxxx"
"""

import asyncio, json, os, re, sys, socket
from datetime import datetime
from playwright.async_api import async_playwright

OUTPUT_DIR = os.path.expanduser("~/.dytranscript_output")
os.makedirs(OUTPUT_DIR, exist_ok=True)
sys.stdout.reconfigure(line_buffering=True)
sys.stderr.reconfigure(line_buffering=True)

def p(msg):
    try: print(msg)
    except UnicodeEncodeError: print(msg.encode("utf-8", errors="replace").decode("gbk", errors="ignore"))

def extract_video_id(text):
    m = re.search(r'douyin\.com/video/(\d+)', text)
    if m: return m.group(1)
    m = re.search(r'v\.douyin\.com/([a-zA-Z0-9_-]+)', text)
    if m: return f"short:{m.group(1)}"
    return None

def fmt_ts(s): m, s = divmod(int(s), 60); return f"{m:02d}:{s:02d}"

def extract(detail):
    r = {"author":(detail.get("author") or {}).get("nickname",""), "desc":detail.get("desc",""),
         "create_time":"", "stats":{}, "hashtags":[], "video_text":[], "has_captions":False}
    s = detail.get("statistics") or {}
    r["stats"] = {"likes":s.get("digg_count",0),"comments":s.get("comment_count",0),"favorites":s.get("collect_count",0)}
    ct = detail.get("create_time",0)
    if ct: r["create_time"] = datetime.fromtimestamp(ct).strftime("%m-%d %H:%M")
    for te in (detail.get("text_extra") or []):
        if te.get("hashtag_name"): r["hashtags"].append(te["hashtag_name"])
    vt = detail.get("video_text") or []
    if vt:
        r["has_captions"] = True
        for seg in vt: r["video_text"].append({"text":seg.get("text",""),"t":fmt_ts(seg.get("start_time",0))})
    return r

def output(result, url):
    p("="*55 + "\n🎬 抖音文案提取完成\n" + "="*55)
    if result["author"]: p(f"👤 {result['author']}")
    if result["create_time"]: p(f"🕐 {result['create_time']}")
    if result.get("stats"):
        s = result["stats"]; p(f"📊 👍{s['likes']}  💬{s['comments']}  ⭐{s['favorites']}")
    if result["hashtags"]: p(f"🏷 {'  '.join('#'+h for h in result['hashtags'])}")
    p(f"🔗 {url}")
    desc = result["desc"].strip()
    if desc: p(f"\n📝 文案描述:\n{'─'*40}\n{desc}")
    vt = result["video_text"]
    if vt:
        p(f"\n🎤 逐字稿 ({len(vt)} 条):\n{'─'*40}")
        ft = []
        for seg in vt: p(f"  [{seg['t']}] {seg['text']}"); ft.append(seg["text"])
        p(f"{'─'*40}\n\n📋 全文:\n{"".join(ft)}")
    elif not result["has_captions"]: p("\n💡 该视频没有自动字幕，只有文案描述。")
    p("="*55)
    ts = datetime.now().strftime("%Y%m%d_%H%M%S")
    nm = re.sub(r'[\\/:*?"<>|]', '', result.get("author","dy"))[:16]
    fp = os.path.join(OUTPUT_DIR, f"dy_{nm}_{ts}.txt")
    with open(fp,"w",encoding="utf-8") as f:
        f.write(f"作者: {result['author']}\n链接: {url}\n\n")
        if desc: f.write(f"文案:\n{desc}\n\n")
        if vt:
            f.write("逐字稿:\n")
            for seg in vt: f.write(f"[{seg['t']}] {seg['text']}\n")
    p(f"\n💾 已保存: {fp}")

async def main():
    if len(sys.argv) < 2:
        p("用法:\n  python dytranscript.py <抖音链接>\n\n首次使用:\n  1. 关闭所有 Chrome\n  2. Win+R 输入: chrome.exe --remote-debugging-port=9222\n  3. 登录抖音\n  4. 运行本脚本")
        return
    input_text = " ".join(sys.argv[1:])
    video_id = extract_video_id(input_text)
    if not video_id: p("❌ 未识别到抖音链接"); return

    video_url = f"https://v.douyin.com/{video_id.split(':')[1]}/" if video_id.startswith("short:") else f"https://www.douyin.com/video/{video_id}"

    p("🚀 抖音逐字稿提取器\n" + f"🕐 {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n🎯 {video_url}")

    # 检查端口
    port_ok = False
    for i in range(3):
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.settimeout(1); r = s.connect_ex(('127.0.0.1', 9222)); s.close()
            if r == 0: port_ok = True; break
        except: pass
        await asyncio.sleep(1)

    if not port_ok:
        p("\n" + "!" * 50)
        p("需要先以调试模式启动 Chrome")
        p("!" * 50)
        p("\n请按以下步骤操作：")
        p("  ① 完全关闭所有 Chrome 窗口")
        p("  ② 按 Win+R，粘贴并回车：")
        p("     chrome.exe --remote-debugging-port=9222")
        p("  ③ 在打开的 Chrome 中登录抖音（如未登录）")
        p("  ④ 重新运行本命令即可")
        p("\n完成后直接重新运行即可，不需要再做其他操作。")
        return

    async with async_playwright() as pw:
        browser = await pw.chromium.connect_over_cdp("http://127.0.0.1:9222")
        ctx = browser.contexts[0] if browser.contexts else await browser.new_context()
        page = ctx.pages[0] if ctx.pages else await ctx.new_page()

        p("\n检查登录...")
        await page.goto("https://www.douyin.com/", wait_until="domcontentloaded", timeout=20000)
        await page.wait_for_timeout(2000)

        cookies = await ctx.cookies()
        logged_in = any(c.get("name") in ("sessionid","sid_guard","sid_tt") and c.get("value") for c in cookies)
        if not logged_in:
            p("⚠ 请在 Chrome 中登录抖音，然后重新运行本命令")
            return
        if not logged_in: p("❌ 未登录"); return
        p("✅ 已登录")

        # 短链解析
        if video_id.startswith("short:"):
            np = await ctx.new_page()
            await np.goto(video_url, wait_until="domcontentloaded", timeout=30000)
            m = re.search(r'douyin\.com/video/(\d+)', np.url)
            if m: video_id = m.group(1); video_url = np.url; p(f"  → 完整 ID: {video_id}")
            await np.close()

        # 提取
        detail_data = {}
        async def on_resp(resp):
            try:
                if "aweme/v1/web/aweme/detail/" in resp.url:
                    data = await resp.json()
                    if data.get("aweme_detail"): detail_data["detail"] = data["aweme_detail"]
            except: pass
        page.on("response", on_resp)
        p("\n打开视频页面...")
        await page.goto(video_url, wait_until="domcontentloaded", timeout=30000)
        for _ in range(20):
            await page.wait_for_timeout(1000)
            if detail_data.get("detail"): break
        await page.wait_for_timeout(3000)

        if detail_data.get("detail"):
            output(extract(detail_data["detail"]), video_url)
        else:
            p("⚠ 未获取到视频数据（可能需要登录或视频不存在）")

        p("\n✅ 完成！Chrome 保持打开即可。")

if __name__ == "__main__":
    try: asyncio.run(main())
    except KeyboardInterrupt: p("\n已取消")
    except Exception as e:
        p(f"\n❌ 错误: {e}")
        import traceback; traceback.print_exc()
